Design/Crystallography/HKL.md

```diff
@@ -0,0 +1,121 @@
+
+While it is a big task to make each and every use of `V3D` (and connected to that also `Kernel::Matrix`) type-safe, implementing a dedicated type for Miller indices HKL is a rather confined task. Potential replacements for `V3D` in this domain can exist in parallel at first and code can be ported step-by-step, simply reverting to the current "unsafe" implementation where necessary.
+
+## Requirements
+
+There should be a datatype different from `V3D` but with similar capabilities that is used to represent Miller index triplets.
+
+While in theory, Miller indices are defined to be composed of three integer numbers, some flexibility is required for cases in early steps of data processing where exact integers may not be known or for commensurate superstructures where fractional indices can occur.
+
+Though similar, those three cases cover different uses and should be covered by different types:
+
+  1. `HKL` (integer, "proper Miller indices")
+  2. `ProHKL` (double, "early stages of reduction")
+  3. `FractionalHKL` (double, use for magnetic or other super-structures)
+
+These types should be by default incompatible, i.e. it should not be possible to add a `ProHKL` onto a `FractionalHKL`-object unless either is explicitly converted to the other. The same is true for comparisons and so on, but as assumed above, it should be possible to perform certain explicit conversions between the types.
+
+Furthermore there are some algorithms or operations that are agnostic to the type, so there should be a way to define algorithms that work for general objects that are "HKL-like".
+
+
+## Design
+
+### IsHKL-mixin
+
+All of the above requirements can be solved by the so called "Curiously Recurring Template Pattern" (CRTP). For this particular purpose, the base class needs to be templated on the numeric type as well as on the derived type as it is normal for the CRTP:
+
+```
+template <typename NumericType, typename Derived>
+class IsHKL {
+public:
+  const NumericType &h() const { ... }
+  void setH(const NumericType &h) { ... }
+  
+  bool operator==(const Derived &rhs) {
+}
+};
+
+class ProHKL : IsHKL<double, ProHKL> {
+};
+
+class FractionalHKL : IsHKL<double, FractionalHKL> {
+};
```

Comment:
I suppose they could, but the underlying storage type is completely hidden away (except kind of for the iterators, but those could be provided in a different way I suppose), so this can (and should) be swapped out later. I would consider that a separate issue.